planner/cascades/transformation_rules.go

```diff
@@ -1611,3 +1613,143 @@ func pushLimitDownOuterJoinToChild(limit *plannercore.LogicalLimit, outerGroup *
+	join := joinExpr.ExprNode.(*plannercore.LogicalJoin)
+
+	var innerChildIdx int
+	switch join.JoinType {
+	case plannercore.LeftOuterJoin:
+		innerChildIdx = 1
+	case plannercore.RightOuterJoin:
+		innerChildIdx = 0
+	default:
+		return nil, false, false, nil
+	}
+	outerGroup := joinExpr.Children[1^innerChildIdx]
+	innerGroup := joinExpr.Children[innerChildIdx]
+
+	outerUniqueIDs := set.NewInt64Set()
+	for _, outerCol := range outerGroup.Prop.Schema.Columns {
+		outerUniqueIDs.Insert(outerCol.UniqueID)
+	}
+	// only when agg only use the columns from outer table can eliminate outer join.
+	if !plannercore.IsColsAllFromOuterTable(agg.GetUsedCols(), outerUniqueIDs) {
+		return nil, false, false, nil
+	}
+	// outer join elimination with duplicate agnostic aggregate functions.
+	_, aggCols := plannercore.GetDupAgnosticAggCols(agg, nil)
+	if len(aggCols) > 0 {
+		newAggExpr := r.doEliminateOuterJoin(agg, join, joinExpr.Schema(), outerGroup, innerChildIdx)
+		return []*memo.GroupExpr{newAggExpr}, true, false, nil
+	}
+	// outer join elimination without duplicate agnostic aggregate functions.
+	innerJoinKeys := join.ExtractJoinKeys(innerChildIdx)
+	contain, err := isInnerJoinKeysContainUniqueKey(innerGroup, innerJoinKeys)
+	if err != nil {
+		return nil, false, false, err
+	}
+	if contain {
+		newAggExpr := r.doEliminateOuterJoin(agg, join, joinExpr.Schema(), outerGroup, innerChildIdx)
+		return []*memo.GroupExpr{newAggExpr}, true, false, nil
+	}
+
+	return nil, false, false, nil
+}
+
+// It will transform `Aggregation->Join->outer,inner` to `Aggregation->outer` or `Aggregation->Selection->outer`.
+func (r *EliminateOuterJoinBelowAggregation) doEliminateOuterJoin(agg *plannercore.LogicalAggregation, join *plannercore.LogicalJoin, joinSchema *expression.Schema, childGroup *memo.Group, innerChildIdx int) *memo.GroupExpr {
+	newAggExpr := memo.NewGroupExpr(agg)
+	ok, sel := convertOuterJoinToSel(join, innerChildIdx)
+	if ok {
+		selExpr := memo.NewGroupExpr(sel)
+		selExpr.SetChildren(childGroup)
+		selGroup := memo.NewGroupWithSchema(selExpr, joinSchema)
+		childGroup = selGroup
+	}
+	newAggExpr.SetChildren(childGroup)
+	return newAggExpr
+}
+
+// Return true only if all the aggregate functions are duplicate agnostic.
+// Only the following functions are considered to be duplicate agnostic:
+//   1. MAX(arg)
+//   2. MIN(arg)
+//   3. FIRST_ROW(arg)
+//   4. Other agg functions with DISTINCT flag, like SUM(DISTINCT arg)
+func (r *EliminateOuterJoinBelowAggregation) isAggFuncAllDupAgnostic(agg *plannercore.LogicalAggregation) bool {
+	for _, aggDesc := range agg.AggFuncs {
+		if !aggDesc.HasDistinct &&
+			aggDesc.Name != ast.AggFuncFirstRow &&
+			aggDesc.Name != ast.AggFuncMax &&
+			aggDesc.Name != ast.AggFuncMin {
+			return false
+		}
+	}
+	return true
+}
+
+// convert outer join to selection when len(outerConditions) > 0.
+func convertOuterJoinToSel(join *plannercore.LogicalJoin, innerChildIdx int) (bool, *plannercore.LogicalSelection) {
```

Comment:
I made a mistake.
About `on clause`, only the `innerConditions` can be pushed down, be converted to `Selection`.

The runtime error because of `selGroup := memo.NewGroupWithSchema(selExpr, joinSchema)`.
The correct one is `selGroup := memo.NewGroupWithSchema(selExpr, childGroup.Prop.Schema)`

Reply:
I test the query `select max(a.a) from t a left join t b on a.a = b.a`  without my change and got an runtime.

panic log
```
PANIC: integration_test.go:150: testIntegrationSuite.TestAggregation

... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x432FCF)

G:/Go/src/runtime/panic.go:679
  in gopanic
G:/Go/src/runtime/panic.go:199
  in panicmem
G:/Go/src/runtime/signal_windows.go:233
  in sigpanic
F:/WareHouse/tidb/planner/core/exhaust_physical_plans.go:162
  in LogicalJoin.GetMergeJoin
implementation_rules.go:510
  in ImplMergeJoin.OnImplement
optimize.go:346
  in Optimizer.implGroupExpr
optimize.go:285
  in Optimizer.implGroup
optimize.go:292
  in Optimizer.implGroup
optimize.go:292
  in Optimizer.implGroup
optimize.go:251
  in Optimizer.onPhaseImplementation
optimize.go:115
  in Optimizer.FindBestPlan
F:/WareHouse/tidb/planner/optimize.go:169
  in optimize
F:/WareHouse/tidb/planner/optimize.go:60
  in Optimize
F:/WareHouse/tidb/planner/core/planbuilder.go:2711
  in PlanBuilder.buildExplain
F:/WareHouse/tidb/planner/core/planbuilder.go:387
  in PlanBuilder.Build
F:/WareHouse/tidb/planner/optimize.go:133
  in optimize
F:/WareHouse/tidb/planner/optimize.go:60
  in Optimize
F:/WareHouse/tidb/executor/compiler.go:61
  in Compiler.Compile
F:/WareHouse/tidb/session/session.go:1112
  in session.execute
F:/WareHouse/tidb/session/session.go:1066
  in session.Execute
F:/WareHouse/tidb/util/testkit/testkit.go:144
  in TestKit.Exec
F:/WareHouse/tidb/util/testkit/testkit.go:235
  in TestKit.MustQuery
integration_test.go:169
  in testIntegrationSuite.TestAggregation
G:/Go/src/reflect/value.go:460
  in Value.call
G:/Go/src/reflect/value.go:321
  in Value.Call
D:/GOPATH/pkg/mod/github.com/pingcap/check@v0.0.0-20191216031241-8a5a85928f12/check.go:850
  in suiteRunner.forkTest.func1
D:/GOPATH/pkg/mod/github.com/pingcap/check@v0.0.0-20191216031241-8a5a85928f12/check.go:739
  in suiteRunner.forkCall.func1
G:/Go/src/runtime/asm_amd64.s:1357
  in goexit
```
May be is because of `join.GetMergeJoin(reqProp, expr.Schema())` in `ImplMergeJoin.OnImplement`.
`join.children` be used in `join.GetMergeJoin` and `join.children` maybe is nil in cascades.

Reply:
Could you fix this panic in this PR, if it is easy to do?

Reply:
ok